umbraco/Umbraco-CMS
# New Backoffice: Health check controller refactor

## Details
- Refactoring the endpoints working with health checks so that their responses contain just the necessary information;
- Splitting the information about health checks into several smaller endpoints;
- The currently available endpoints are:
  - GET `/health-check-group` - gets all health check group names in a paginated list;

> Response
```json
{
  "total": 0,
  "items": [
    {
      "name": "string"
    }
  ]
}
```
-
  - GET `/health-check-group/{name} ` - gets a health check group found with the same name, together with all its health checks or Not Found response;

> Response
```json
{
  "name": "string",
  "checks": [
    {
      "key": "3fa85f64-5717-4562-b3fc-2c963f66afa6",
      "name": "string",
      "description": "string"
    }
  ]
}
```

-
  - POST `/health-check-group/{name}/check` - check all health checks in the group with a given group name or Not Found response;

> Response
```json
{
  "checks": [
    {
      "key": "3fa85f64-5717-4562-b3fc-2c963f66afa6",
      "results": [
        {
          "message": "string",
          "resultType": "Success",
          "actions": [
            {
              "healthCheckKey": "3fa85f64-5717-4562-b3fc-2c963f66afa6",
              "alias": "string",
              "name": "string",
              "description": "string",
              "valueRequired": true,
              "providedValue": "string",
              "providedValueValidation": "string",
              "providedValueValidationRegex": "string"
            }
          ],
          "readMoreLink": "string"
        }
      ]
    }
  ]
}
```

-
  - `health-check/execute-action` (_same as before_)- executes a given action from a health check or retuns a _Bad Request_;
    - If a health check provides actions that can be executed to fix a health check when it fails, then we can use the endpoint to rectify the issue;
    - The **request body** is in the form of a HealthCheckAction:
```json
{
  "healthCheckKey": "3fa85f64-5717-4562-b3fc-2c963f66afa6",
  "alias": "string",
  "name": "string",
  "description": "string",
  "valueRequired": true,
  "providedValue": "string",
  "providedValueValidation": "string",
  "providedValueValidationRegex": "string"
}
```
</br>


## Test
- Check the responses with the information received from the backoffice:
  - `/health-check-group` - _GET_ request
      - Execute with the default skip and take values or modify them as you like;
      - Check the response with the health check groups that we get in the backoffice - we should have the same amount of groups;
      - Try adding a custom health check, as shown in the code example of [general checks section](https://docs.umbraco.com/umbraco-cms/extending/health-check#general-checks) and make sure "_SEO_" group is included in the response;
      - Try to disable a health check in _appsettings.json_ and verify that if this is the only check in a group, that group name is NOT included in the response;
        - You can get their ids/keys from _[here](https://docs.umbraco.com/umbraco-cms/extending/health-check#built-in-checks)_;
        - _Note:_ You have to disable all checks from a group not to include it in the response.

_ex:_
```json
"Umbraco": {
  "CMS": {
    "HealthChecks": {
      "DisabledChecks": [
        {
          "Id": "73DD0C1C-E0CA-4C31-9564-1DCA509788AF" // disables the only "Data Integrity" health check
        }
      ]
    }
  }
}
```
  
- 
  - `/health-check-group/{name}` - _GET_ request
    - Edit the `name` - ex: `Permissions`;  
    - Make sure the details of the group are correct (_you can compare them to the response from the old endpoint `umbraco/backoffice/umbracoapi/healthcheck/GetAllHealthChecks`_);
    - Pass a group name that doesn't exist - ex: _"Data Integrity"_ since we disabled all its health checks and verify that you get a is Not Found response.
  - `/health-check-group/{name}/check` - _POST_ request
    - Pass the group `name` that you want to execute the checks for - ex: `Permissions`;  
    - Make sure that the information in the result(s) (`"results": []`) is identical to the ones from the backoffice (_see the response of `/umbraco/backoffice/umbracoapi/healthcheck/GetStatus?id=53dba282-4a79-4b67-b958-b29ec40fcc23`_);
    - Pass a group name that doesn't exist - ex: _"Data Integrity"_ since we disabled all its health checks and verify that you get a is Not Found response.

  - `health-check/execute-action` - _POST_ request (_nothing changed for this one_)
    - You can add the custom health check shown in the code example (the whole `RobotsTxt` class) in _[general checks section](https://docs.umbraco.com/umbraco-cms/extending/health-check#general-checks)_;
      - It adds an action that you can execute, which will add a robots.txt file and therefore satisfy the health check;
    - Check the returned response from clicking the action button of the SEO > Robots.txt healthcheck in the backoffice (it will make a call to the old ExecuteAction endpoint);
    - Delete the created robots.txt file;
    - Go to `localhost/umbraco/swagger` and try the `health-check/execute-action` endpoint with the following body:
```json
{
  "healthCheckKey": "3a482719-3d90-4bc1-b9f8-910cd9cf5b32",
  "alias": "addDefaultRobotsTxtFile",
  "name": "Do the thing!",
  "description": "Clicking button will do the thing and fix your site",
  "valueRequired": false,
  "providedValue": null,
  "providedValueValidation": null,
  "providedValueValidationRegex": null
}
```
-
  -
    - Compare the response body with the one from the backoffice;
    - Verify that the response from `/health-check-group/SEO` no longer contains any actions and that the `resultType` changed from `"Error"` to `"Success"`.
